feat: first commit of activities

cdoak-gh opened the pull request:
Quickbase is a new implementation partner that provides a SaaS work management platform. Te platform includes a REST API which we used to integrate Quickbase records with VertiGIS Studio via Workflow. There are a number of professional services projects in the pipeline and these implementations will be greatly improved by a stable and maintained activity pack.

cdoak-gh committed:
feat: first commit of activities

kenlyon commented on README.md:
```diff
@@ -31,12 +27,11 @@ To use these activities in [VertiGIS Studio Workflow Designer](https://apps.vert
-The quickbase connection requires a valid user token in order.  It is the responsibility of the Workflow implmentor to retreive and store the token.  [Click here](https://developer.quickbase.com/auth) for more information on how to generate a Quickbase user token.
-    1. Add the `Create App Service` activity to a workflow
-    1. Set the `URL` input to the root URL the Quickbase service. For example, `https://api.quickbase.com/v1`.
+The Quickbase activity pack is intended to be used exclusively by VertiGIS Studio Web applications embedded in browser window within Quickbase, and a secure postMessage communication channel is established with Quickbase using postMessage.  Once the communication is established, the Workflow will request tokens using the identity of the user logged into Quickbase.
```
If this activity pack is exclusively for VSW then I would recommend we remove `EXB` from the `@supportedApps` tag in all of the activities.

cdoak-gh replied:
Fixed.

kenlyon commented on README.md:
```diff
@@ -31,12 +27,11 @@ To use these activities in [VertiGIS Studio Workflow Designer](https://apps.vert
-The quickbase connection requires a valid user token in order.  It is the responsibility of the Workflow implmentor to retreive and store the token.  [Click here](https://developer.quickbase.com/auth) for more information on how to generate a Quickbase user token.
-    1. Add the `Create App Service` activity to a workflow
-    1. Set the `URL` input to the root URL the Quickbase service. For example, `https://api.quickbase.com/v1`.
+The Quickbase activity pack is intended to be used exclusively by VertiGIS Studio Web applications embedded in browser window within Quickbase, and a secure postMessage communication channel is established with Quickbase using postMessage.  Once the communication is established, the Workflow will request tokens using the identity of the user logged into Quickbase.
```
> embedded in browser window

We're missing a word here. "embedded in a browser window"?

cdoak-gh replied:
Fixed

kenlyon commented on src/activities/CreateQuickbaseService.ts:
```diff
@@ -0,0 +1,53 @@
+    if (!hostName) {
+      throw new Error("hostName is required");
+    }
+    const quickbaseUri = quickbaseUrl.replace(/\/*$/, "");
```
It's a small detail, but this regular expression will still match when there's no `/` at the end of the url. It'll replace `""` with `""`.

I doubt there's much of a performance hit with what you have, but `/\/+$/` would only match if there was one or more `/` at the end, which I think is the intended behaviour.

cdoak-gh replied:
Fixed

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+
+  constructor(url, hostname, origin) {
+    this.hostName = hostname;
+    this.instanceUrl = url;
```
Would it be tidier to clean up the trailing `/` from the url here?

cdoak-gh replied:
Yeah. That makes sense.  I moved the check.

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+export default class QuickbaseService {
+  instanceUrl: string;
+  hostName: string;
+  _origin: string;
```
Should this be `private`?

cdoak-gh replied:
Yes. Fixed.

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+  _origin: string;
+  private _tokens = new Map<string, QuickbaseToken>();
+
+  constructor(url, hostname, origin) {
```
Missing type declarations on the constructor parameters.

cdoak-gh replied:
fixed

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+  async getToken(id: string): Promise<string> {
+    const qbToken = this._tokens.get(id);
+    if (qbToken && qbToken.expiration < Date.now()) {
+      return `QB-TEMP-TOKEN ${qbToken.token}`;;
```
Extra `;` here.

cdoak-gh replied:
fixed

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+      const message = await this.postMessageAwaitReply(id);
+      if (message && message.error) {
+        throw new Error(message.error)
+      } else if (message.parameters && message.parameters.token) {
```
You would get an error here if `message` was `undefined`.

cdoak-gh replied:
Fixed the message check for undefined.

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+      );
+    }
+
+    if (!this.hostName) {
```
Should this be validated in the constructor? I can't see any valid use case where you'd create an instance of the service and not call this method.

cdoak-gh replied:
I agree. Moved it to the constructor.

kenlyon replied:
I see the owner and url are now checked in the constructor, but the host is still not validated. Is this an oversight?

cdoak-gh replied:
Yes. Fixed.

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+
+
+export default class QuickbaseService {
+  instanceUrl: string;
```
Should this be `readonly` to ensure it's not changed after creation?

cdoak-gh replied:
Good idea. Fixed.

kenlyon commented on src/QuickbaseService.ts:
```diff
@@ -0,0 +1,67 @@
+
+export default class QuickbaseService {
+  instanceUrl: string;
+  hostName: string;
```
Should this be `readonly` to ensure it's not changed after creation?

cdoak-gh replied:
Yup. Fixed.

kenlyon commented:
I notice an npm audit failure when running `npm install`. I think it's an issue with the Workflow SDK, though. I should create an updated version.

kenlyon commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
```
(This comment is not related to this file, but I couldn't see a way to add a review comment outside of the changed files.)

I notice an npm audit failure when running `npm install`. If you run `npm audit fix` it will fix it.

cdoak-gh replied:
Ran it and updated the package-lock.json

cdoak-gh committed:
fix:
- removed exb from supported apps
- corrected typo in READ.ME
- QuickbaseService:
  - moved quickbaseUrl from CreateQuickbaseService
  - fixed regular expression for cleaning quickbaseUrl
  - set owner window as private global and moved validation to constructor
  - made origin private
  - added type declarations on the constructor parameters.
  - fixed message checking logic
  - made instanceUrl readonly
- package-lock.json:
  - ran npm audit fix

cdoak-gh committed:
fix:
- QuickbaseService
 - set hostname as readonly

cdoak-gh committed:
fix:
QuickbaseService:
 - validate hostname in constructor
